### PR TITLE
[#29] Final points

### LINKED ISSUE #29: Refine/improve documentation
# Clarification and motivation

Currently we don't have enough documentation. We have:
1. Auto generated documentation referred from `README.md`. It's quite small and doesn't even describe entrypoints.
2. `docs/specification.md` which describes what contract needs to be implemented, but doesn't really explain how to use it.
3. Trivial build and usage instructions.

# Acceptance criteria

1. Auto generated documentation is expanded and contains as much as possible. For example: storage description, description of each entrypoint.
2. There are detailed usage instructions. In particular, they should explain how to instantiate parameters of the contract. The contract we developed is a base contract for various DAOs, we should clearly explain how to implement DAOs.
3. `docs/specification.md` is proofread to match our code. Well, it may be too hard to check _everything_ carefully, but at least let's check that error scenarios and corresponding errors are accurate.

### PR DESCRIPTION
**Pull request by Martoon-00**
## Description

Some last changes related to #29.

## Related issue(s)

Resolves #29

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the following policy](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

**Comment by Martoon-00** [comment]
It seems to me that I depend on a commit in hackage.nix which is further than the merged fix, but the issue is here. What am I doing wrong :thinking:

**Comment by zhenyavinogradov** [comment]
The problem is that `sha256` for haskell.nix is from the old pin, where the snapshot wasn't working, so nix tries to use the cached version of haskell.nix instead of downloading the new one. But anyway, I'd keep using our fork for now, because it's based on the same commit that most of our other projects are using, so it would avoid rebuilding ghc and some libraries. At some point we'll update haskell.nix over all repos at once, and we'll repin it to upstream here as well.

**Comment by Martoon-00** [comment]
Ah okay, I somehow thought sha256 is updated after I do `niv update hackage.nix`.

Anyway, I'll remove that commit for now.

**Comment by gromakovsky** [comment]
> But anyway, I'd keep using our fork for now, because it's based on the same commit that most of our other projects are using, so it would avoid rebuilding ghc and some libraries.

Do you mean that the diff between `haskell.nix` revision used here and the one used in most of our repos is much smaller than between the latest upstream master version and the one used in most of our repos? And in particular this first diff doesn't include ghc and some libraries (i. e. they don't differ), but the second diff does include them?

**Comment by gromakovsky** [comment]
> FA2 metadata handling is currently not implemented, but people ask questions about metadata. Ideally we'd like to do #42 first, i. e. implement TZIP-16 metadata from the beginning rather than implement the current FA2 way (which will not be recommended soon).

Oh, meh, apparently I forgot to finish this item or some part of it was lost :facepalm: 

I wanted to propose to write explicitly (e. g. in `specification.md`) that metadata handling is currently in some intermediate state and will be updated when https://gitlab.com/tzip/tzip/-/merge_requests/94 is merged.

> Resolves (partially) #29

Why partially this time? It kinda contradicts your "final" and "last" :)
Is it because you want to wait for `haskell.nix` update? If there is a good reason not to do it now, I think we can forget about it until someone creates an issue about updating revisions in all our repos.
Or because I wrote some vague "Ideally we'd like to do #42 first"? It was an unfinished item :)

**Comment by Martoon-00** [comment]
> Why partially this time? It kinda contradicts your "final" and "last" :)

I always could create `Final final points` :smile: :pled:.

File: docs/specification.md [hunk 1]
```diff
@@ -831,3 +829,9 @@ nat
 ```
 
 - For `vote` entrypoint with permit, returns the current suitable counter for constructing permit signature.
+
+# TZIP-16 metadata
```
Comment (gromakovsky) [inline docs/specification.md:833]: I suppose it should be added to ToC.

**Review by gromakovsky — APPROVED** [review]
Ok, one final request.

**Comment by zhenyavinogradov** [comment]
@Martoon-00 `niv` usually updates both revision and sha256 automatically, I'm not sure why it didn't.

@gromakovsky Yes, the diff between `haskell.nix` revision used in other repos and the branch used here contains just the fix for the snapshot path and no other changes, so this repo will use the same ghc build.

**Pull request merged**